{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "This script to process all Sanger sequencing files from an Intein-assisted Bisection Mapping (IBM) experiment within a directory and return the split sites where the target protein was split and the split intein was inserted.\n",
    "  \n",
    "![IBM Sanger Overview](images/Github_Sanger_overview.png)\n",
    "  \n",
    "In brief, a \"signature sequence\" that marks the end of C-lobe intein is aligned to the PCR product to identify a short sequence immediately downstream, which is extracted. The extracted sequence is then aligned to the intact target protein CDS sequence to identify the DNA position where the CDS was split. The -1 and +1 DNA positions are then used to infer where the protein sequence was split.\n",
    "\n",
    "## Dependences\n",
    "1. Biopython\n",
    "2. Pandas\n",
    "  \n",
    "## Input\n",
    "1. **.ab1 files under a subdirectory**  \n",
    "Data from Source BioScience (SBS) has Sequence Record ID of  \n",
    "`459313601_IBMs0768_IBMo0858_B09.ab1`,  \n",
    "which is   \n",
    "`<order no.>_<sample name or ID>_<primer name>_<well used in SBS>.ab1`  \n",
    "  \n",
    "  If your sequence record ID is different you will need to customize the script below.\n",
    "    \n",
    "  The subdirectory in this case is `./IBM_seq_example/`.\n",
    "\n",
    "  \n",
    "2. **target CDS sequence**  \n",
    "    Take the 1st bp of CDS to ~100-200 bp after stop codon\n",
    "  \n",
    "  \n",
    "3. **signature sequence**\n",
    "    Take 30 - 50 bp of sequence from the C-lobe intein at the C terminal, refer to figure above.\n",
    "\n",
    "## Output\n",
    "A csv file with split sites mapped for each sequencing read"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO, Align\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_split_site(folderDir, signature_sequence, target_CDS_seq):\n",
    "\n",
    "    ss_output = pd.DataFrame()  # set up empty df to store all data\n",
    "\n",
    "    # Parse all files in the same folder\n",
    "\n",
    "    for file in os.listdir(folderDir):\n",
    "\n",
    "        filename = os.fsdecode(file)\n",
    "\n",
    "        if filename.endswith(\".ab1\"):\n",
    "            ## Import the data & extract metadata from the sequenicng result\n",
    "            seqDir = os.path.join(folderDir,filename)\n",
    "            record = SeqIO.read(seqDir, \"abi\")\n",
    "            seq_order,strainID,seq_Primer = record.id.split('_')\n",
    "\n",
    "            ## Sequence alignment part\n",
    "\n",
    "            # Set up aligner for alignment\n",
    "            aligner = Align.PairwiseAligner()\n",
    "            aligner.mode = 'local'\n",
    "            aligner.open_gap_score = -10\n",
    "            aligner.extend_gap_score = -10\n",
    "\n",
    "            # Reset information to be written\n",
    "            dna_ss_minus1 = 'n/a'\n",
    "            dna_ss_plus1 = 'n/a'\n",
    "            aa_ss_minus1 = 'n/a'\n",
    "            aa_ss_plus1 = 'n/a'\n",
    "            aa_ss_middle = 'n/a'\n",
    "\n",
    "            # Perform local alignment between the sequencing result and the signature sequence\n",
    "            alignment = aligner.align(record.seq,signature_sequence)\n",
    "            # Locate the indices where the signature sequence was aligned to\n",
    "            if alignment: # Analyze only those where alignment is found\n",
    "                sig_start,sig_end = alignment[0].path[0][0], alignment[0].path[1][0]    \n",
    "                # Extract the bp that would correspond to the target CDS\n",
    "                extracted_seq = record.seq[sig_end:]   # Note: no need to +1 here because the indexing of strings starts from 0\n",
    "                extracted_seq = extracted_seq[0:31] # Only take the first 30 bp of sequence\n",
    "                # Perform local alignment between the extracted sequence and the target CDS sequence\n",
    "                target_CDS = Seq(target_CDS_seq) # import the target CDS sequence as a Seq.Seq object\n",
    "                alignment = aligner.align(target_CDS,extracted_seq)\n",
    "                # Locate the start index of where the extracted sequence was aligned to, i.e. ID the split site\n",
    "                if alignment:\n",
    "                    # problem: a poor alignemnt of signature sequence might return abberent alignment at this stage as well\n",
    "                    \n",
    "                    dna_ss_minus1 = alignment[0].path[0][0] # Note: indexing of string starts from 0, so this becomes -1\n",
    "                    # Workout the rest of the split sites\n",
    "                    dna_ss_plus1 = dna_ss_minus1 + 1\n",
    "                    aa_ss_minus1 = dna_ss_minus1 / 3\n",
    "                    aa_ss_plus1 = aa_ss_minus1 + 1\n",
    "                    aa_ss_middle = aa_ss_minus1 + 0.5\n",
    "\n",
    "            # Create a dictionary for import into pandas dataframe\n",
    "            ss_data = {\n",
    "                    'order': seq_order,\n",
    "                    'primer': seq_Primer,\n",
    "                    'raw_seq': record.seq._data, # raw sequence\n",
    "                    'read_length': len(record.seq), # length of the raw sequence\n",
    "                    'dna_ss_minus1': dna_ss_minus1,\n",
    "                    'dna_ss_plus1':  dna_ss_plus1,\n",
    "                    'aa_ss_minus1': aa_ss_minus1,\n",
    "                    'aa_ss_plus1': aa_ss_plus1,\n",
    "                    'aa_ss_middle':  aa_ss_middle # split site (e.g. 196.5)\n",
    "                    }\n",
    "\n",
    "            ss_dfRow = pd.DataFrame(ss_data,index=[strainID])\n",
    "            ss_output = ss_output.append(ss_dfRow, sort=True)\n",
    "            \n",
    "    return ss_output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Execution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Provide the CDS of the target protein\n",
    "target_CDS_seq = 'atgagtattcaacatttccgtgtcgcccttattcccttttttgcggcattttgccttcctgtttttgctcacccagaaacgctggtgaaagtaaaagatgctgaagatcagttgggtgcacgagtgggtTACatcgaactggatctcaacagcggtaagatccttgagagttttcgccccgaagaacgttttccaatgatgagcacttttaaagttctgctatgtggcgcggtattatcccgtattgacgccgggcaagagcaactcggtcgccgcatacactattctcagaatgacttggttgagtactcaccagtcacagaaaagcatcttacggatggcatgacagtaagagaattatgcagtgctgccataaccatgagtgataacactgcggccaacttacttctgacaacgatcggaggaccgaaggagctaaccgcttttttgcacaacatgggggatcatgtaactcgccttgatcgttgggaaccggagctgaatgaagccataccaaacgacgagcgtgacaccacgatgcctgtagcaatggcaacaacgttgcgcaaactattaactggcgaactacttactctagcttcccggcaacaattaatagactggatggaggcggataaagttgcaggaccacttctgcgctcggcccttccggctggctggtttattgctgataaatctggagccggtgagcgtGGAtctcgcggtatcattgcagcactggggccagatggtaagccctcccgtatcgtagttatctacacgacggggagtcaggcaactatggatgaacgaaatagacagatcgctgagataggtgcctcactgattaagcattggTAAtactagagaacgcatgagAAAGCCCCCGGAAGATCACCTTCCGGGGGCTTTtttattgcgctactagtagcggccgctgcaggagtcactaagggttagttagttagat'\n",
    "# CDS for beta-lactamase\n",
    "\n",
    "# Provide the signature sequence\n",
    "signature_sequence= 'CCTGTTCTATGCCAATGATATTCTGACCCATAATTCGTCA' # signature sequence for gp41-1 intein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa_ss_middle</th>\n",
       "      <th>aa_ss_minus1</th>\n",
       "      <th>aa_ss_plus1</th>\n",
       "      <th>dna_ss_minus1</th>\n",
       "      <th>dna_ss_plus1</th>\n",
       "      <th>order</th>\n",
       "      <th>primer</th>\n",
       "      <th>raw_seq</th>\n",
       "      <th>read_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IBMs0758</th>\n",
       "      <td>192.5</td>\n",
       "      <td>192.0</td>\n",
       "      <td>193.0</td>\n",
       "      <td>576</td>\n",
       "      <td>577</td>\n",
       "      <td>459313601</td>\n",
       "      <td>IBMo0858</td>\n",
       "      <td>NNNNNNNNNNTTGANTTAGCGGTACCACCTGTTCTATGCCAATGAT...</td>\n",
       "      <td>551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IBMs0793</th>\n",
       "      <td>196.5</td>\n",
       "      <td>196.0</td>\n",
       "      <td>197.0</td>\n",
       "      <td>588</td>\n",
       "      <td>589</td>\n",
       "      <td>459313601</td>\n",
       "      <td>IBMo0858</td>\n",
       "      <td>NNNNNNNNNNNNGAGTTAGCGGTACCACCTGTTCTATGCCAATGAT...</td>\n",
       "      <td>538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IBMs0757</th>\n",
       "      <td>260.5</td>\n",
       "      <td>260.0</td>\n",
       "      <td>261.0</td>\n",
       "      <td>780</td>\n",
       "      <td>781</td>\n",
       "      <td>459313601</td>\n",
       "      <td>IBMo0858</td>\n",
       "      <td>NNNNNNNGNNTTGANTTAGCGGTAACCACCTGTTCTATGCCAATGA...</td>\n",
       "      <td>347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IBMs0753</th>\n",
       "      <td>261.5</td>\n",
       "      <td>261.0</td>\n",
       "      <td>262.0</td>\n",
       "      <td>783</td>\n",
       "      <td>784</td>\n",
       "      <td>459313601</td>\n",
       "      <td>IBMo0858</td>\n",
       "      <td>NNNNNNNNNNTTGANTTAGCGGTACCACCTGTTCTATGCCAATGAT...</td>\n",
       "      <td>345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IBMs0761</th>\n",
       "      <td>264.5</td>\n",
       "      <td>264.0</td>\n",
       "      <td>265.0</td>\n",
       "      <td>792</td>\n",
       "      <td>793</td>\n",
       "      <td>459313601</td>\n",
       "      <td>IBMo0858</td>\n",
       "      <td>NNNNNNNNNTTGAGTTAGCGGTACCACCTGTTCTATGCCAATGATA...</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IBMs0810</th>\n",
       "      <td>267.5</td>\n",
       "      <td>267.0</td>\n",
       "      <td>268.0</td>\n",
       "      <td>801</td>\n",
       "      <td>802</td>\n",
       "      <td>459313601</td>\n",
       "      <td>IBMo0858</td>\n",
       "      <td>NNNNNNNGNNTTGANTTAGCGGNNNCACCTGTTCTATGCCAATGAT...</td>\n",
       "      <td>325</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          aa_ss_middle  aa_ss_minus1  aa_ss_plus1  dna_ss_minus1  \\\n",
       "IBMs0758         192.5         192.0        193.0            576   \n",
       "IBMs0793         196.5         196.0        197.0            588   \n",
       "IBMs0757         260.5         260.0        261.0            780   \n",
       "IBMs0753         261.5         261.0        262.0            783   \n",
       "IBMs0761         264.5         264.0        265.0            792   \n",
       "IBMs0810         267.5         267.0        268.0            801   \n",
       "\n",
       "          dna_ss_plus1      order    primer  \\\n",
       "IBMs0758           577  459313601  IBMo0858   \n",
       "IBMs0793           589  459313601  IBMo0858   \n",
       "IBMs0757           781  459313601  IBMo0858   \n",
       "IBMs0753           784  459313601  IBMo0858   \n",
       "IBMs0761           793  459313601  IBMo0858   \n",
       "IBMs0810           802  459313601  IBMo0858   \n",
       "\n",
       "                                                    raw_seq  read_length  \n",
       "IBMs0758  NNNNNNNNNNTTGANTTAGCGGTACCACCTGTTCTATGCCAATGAT...          551  \n",
       "IBMs0793  NNNNNNNNNNNNGAGTTAGCGGTACCACCTGTTCTATGCCAATGAT...          538  \n",
       "IBMs0757  NNNNNNNGNNTTGANTTAGCGGTAACCACCTGTTCTATGCCAATGA...          347  \n",
       "IBMs0753  NNNNNNNNNNTTGANTTAGCGGTACCACCTGTTCTATGCCAATGAT...          345  \n",
       "IBMs0761  NNNNNNNNNTTGAGTTAGCGGTACCACCTGTTCTATGCCAATGATA...          334  \n",
       "IBMs0810  NNNNNNNGNNTTGANTTAGCGGNNNCACCTGTTCTATGCCAATGAT...          325  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Execute the split site identification function\n",
    "\n",
    "split_site_data = identify_split_site(\"./IBM_seq_example/\", signature_sequence, target_CDS_seq)\n",
    "\n",
    "split_site_data.sort_values(\"aa_ss_middle\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export as a csv file\n",
    "split_site_data.to_csv(\"example_bla_split_sites.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
